src/nvim/channel.c

```diff
@@ -802,6 +802,7 @@ static void term_close(void *data)
   Channel *chan = data;
   process_stop(&chan->stream.proc);
   multiqueue_put(chan->events, term_delayed_free, 1, data);
+  terminal_wipe(chan->term);
```

Comment:
term_delayed_free calls `terminal_destroy`, seems like `chan->term` may not be safe to use after that. Why not do this from `terminal_enter` as before? That also allows us to mark terminal_wipe as `static`.

Reply:
> term_delayed_free calls terminal_destroy, seems like chan->term may not be safe to use after that.

I'll add a non-null check on `term` in `terminal_wipe`.

> Why not do this from terminal_enter as before? That also allows us to mark terminal_wipe as static.

It didn't make sense to me that the function that handles input to a terminal buffer was also responsible for handling how to close the buffer window. Putting it in `term_close` seems both like a more appropriate place for it and also allows us to re-use the behavior of `terminal_wipe` from anywhere we can trigger `term_close` (e.g. we can re-use it for #15440).

Reply:
> also allows us to re-use the behavior of `terminal_wipe`

hmm. the general "close while maintaining split" behavior is not really related to terminals, it is useful for any buffer. Long-standing feature request...  https://github.com/neovim/neovim/issues/2434

Reply:
In that case, should we wait for a solution to #2434 and apply that to terminal buffers? Or should we move forward with this one with an ad-hoc solution for now and update it retroactively with the #2434 solution.

Reply:
The reason I wanted to put this in `term_close` is to avoid duplicate implementations both when the user presses a key on the `[Process exited 0]` message as well as when the buffer automatically closes in #15440. If we keep this in `terminal_enter` then we have to basically do the same thing again in #15440 but in a different way (unless we can make a `TermClose` autocommand that enters the `terminal_enter` code path with something like `feedkeys`, but that feels pretty hacky to me).

Reply:
> Or should we move forward with this one with an ad-hoc solution for now and update it retroactively with the #2434 solution.

We don't need to wait for #2434 but mainly i'm suggesting to at least put the skeleton of whatever you are starting here, in a more generic place. Or just name it more generically: buf_close_keepwin